File: babbleon/vault/tpm.py
"""
TPM2 vault backend: host_secret sealed to PCR measurements.

Seals against PCRs 4 (boot manager code), 8 (kernel cmdline),
9 (initrd) in ADDITION to PCR7 (Secure Boot). PCR7-only sealing
is bypassable via custom initrd with manipulated PCR7 (oddlama technique).

Requires: tpm2-tools in PATH, tpm2-tss library.

REVIEW(manual): authorized policies for post-kernel-update re-seal are
not yet implemented. Current flow requires manual re-seal after kernel
updates. See tpm2-tools `tpm2_policyauthorize` docs for the pattern.

REVIEW(manual): this module executes tpm2_* subprocesses. On real
hardware, test with tpm2-abrmd daemon vs direct /dev/tpm0 access;
resource manager behavior differs between distros.
"""
import hashlib
import json
import os
import pathlib
import subprocess
import tempfile

from ..errors import HardwareUnavailable, VaultError

# PCRs that must be included in the seal policy; do NOT drop 8/9.
_SEAL_PCRS = "4,7,8,9"
_CONTEXT_FILE = pathlib.Path("~/.config/babbleon/tpm-context.ctx").expanduser()


def _run(cmd: list[str], **kwargs) -> subprocess.CompletedProcess:
    try:
        return subprocess.run(cmd, check=True, capture_output=True, **kwargs)
    except FileNotFoundError:
        raise HardwareUnavailable("tpm2-tools not found in PATH")
    except subprocess.CalledProcessError as exc:
        raise VaultError(f"tpm2 command failed: {exc.stderr.decode(errors='replace')}") from exc


class TPMBackend:
    """
    KEK backend: key sealed to TPM2 PCR policy.

    derive_age_passphrase() is called at unlock time; it unseals the
    KEK from the TPM and returns it as the age passphrase.

    The sealed blob (context file) is stored at _CONTEXT_FILE.
    On a fresh machine, call TPMBackend.enroll(kek_bytes) once.
    """

    def __init__(self, context_path: pathlib.Path | None = None) -> None:
        self.context_path = context_path or _CONTEXT_FILE

    @classmethod
    def available(cls) -> bool:
        try:
            subprocess.run(["tpm2_getcap", "properties-fixed"], check=True,
                           capture_output=True, timeout=5)
            return True
        except Exception:
            return False

    def enroll(self, kek_bytes: bytes) -> None:
        """Seal kek_bytes to current PCR values. Run after each kernel update."""
        self.context_path.parent.mkdir(parents=True, exist_ok=True)
        with tempfile.NamedTemporaryFile(delete=False) as f:
            f.write(kek_bytes)
            tmp = f.name
        try:
            _run([
                "tpm2_create", "-C", "e",
                "-i", tmp,
                "-u", str(self.context_path) + ".pub",
                "-r", str(self.context_path) + ".priv",
                "-L", f"pcr:{_SEAL_PCRS}",
            ])
        finally:
            os.unlink(tmp)

    def _unseal(self) -> bytes:
        with tempfile.NamedTemporaryFile(delete=False, suffix=".out") as f:
            out_path = f.name
        try:
            _run([
                "tpm2_unseal",
                "-u", str(self.context_path) + ".pub",
                "-r", str(self.context_path) + ".priv",
                "-L", f"pcr:{_SEAL_PCRS}",
                "-o", out_path,
            ])
            return pathlib.Path(out_path).read_bytes()
        finally:
            if pathlib.Path(out_path).exists():
                os.unlink(out_path)

    def derive_age_passphrase(self, _credential=None) -> str:
        kek_bytes = self._unseal()
        return hashlib.sha256(kek_bytes + b"babbleon-tpm-age-v1").hexdigest()

File: crates/babbleon/src/vault/tpm.rs
//! TPM2-sealed KEK backend.
//!
//! Wraps the age passphrase with a TPM2 key sealed against the boot state
//! (PCRs 4 + 7 + 8 + 9 — not PCR 7 alone; see RESEARCH T5).  An attacker
//! with disk access cannot unseal without booting the same kernel +
//! shim + GRUB.
//!
//! Compiled only with `--features tpm`.  Without `tss-esapi` the module
//! exposes the same API but returns `HardwareUnavailable`, so the rest
//! of the crate doesn't grow `cfg` noise.
//!
//! DEFERRED M2.5:
//!   - `tpm2_policyauthorize` for kernel-update re-seal without manual flow
//!   - `tpm2-abrmd` vs `/dev/tpm0` resource manager matrix
//!   - Test matrix: real TPM hardware

use crate::errors::{BabbleonError, Result};
use crate::vault::backend::KekBackend;

/// PCRs we seal against.  PCR 7 alone is bypassable (oddlama 2023);
/// PCRs 4 (boot manager), 7 (Secure Boot), 8 (kernel cmdline), 9 (initrd)
/// give the layered measurement the bypass attack can't forge.
pub const SEAL_PCRS: &[u32] = &[4, 7, 8, 9];

pub struct TpmBackend;

impl Default for TpmBackend {
    fn default() -> Self {
        TpmBackend
    }
}

impl KekBackend for TpmBackend {
    fn derive_age_passphrase(&self, _credential: Option<&str>) -> Result<String> {
        #[cfg(feature = "tpm")]
        {
            // The real implementation lives behind the `tpm` feature and
            // pulls in `tss-esapi`.  Stubbed for the no-feature build.
            Err(BabbleonError::HardwareUnavailable(
                "tpm backend: tss-esapi wiring lands in M2.5".into(),
            ))
        }
        #[cfg(not(feature = "tpm"))]
        {
            Err(BabbleonError::HardwareUnavailable(
                "tpm backend: rebuild babbleon with --features tpm".into(),
            ))
        }
    }

    fn name(&self) -> &'static str {
        "tpm"
    }
}
